src/components/mc/cuda/mc_cuda.c

```diff
@@ -78,6 +78,66 @@ static ucc_status_t ucc_mc_cuda_mem_free(void *ptr)
     return UCC_OK;
 }
 
+static inline ucc_status_t ucc_mc_cuda_memcpy_kind_map(ucc_memory_type_t dst_mem,
+                                                       ucc_memory_type_t src_mem,
+                                                       cudaMemcpyKind *kind)
+{
+    ucc_status_t status = UCC_OK;
+    switch (dst_mem) {
+    case UCC_MEMORY_TYPE_HOST:
+        ucc_assert(src_mem == UCC_MEMORY_TYPE_CUDA);
+        *kind = cudaMemcpyDeviceToHost;
+        break;
+    case UCC_MEMORY_TYPE_CUDA:
+        ucc_assert(src_mem == UCC_MEMORY_TYPE_CUDA ||
+                   src_mem == UCC_MEMORY_TYPE_HOST);
+        *kind = (src_mem == UCC_MEMORY_TYPE_HOST) ?
+            cudaMemcpyHostToDevice : cudaMemcpyDeviceToDevice;
+        break;
+    default:
+        status = UCC_ERR_INVALID_PARAM;
+        break;
+    }
+    return status;
+}
+
+static ucc_status_t ucc_mc_cuda_memcpy(void *dst, const void *src, size_t len,
+                                       ucc_memory_type_t dst_mem,
+                                       ucc_memory_type_t src_mem)
+{
+    cudaError_t    st;
+    cudaMemcpyKind kind;
+    ucc_status_t   status;
+    ucc_assert(dst_mem == UCC_MEMORY_TYPE_CUDA ||
+               src_mem == UCC_MEMORY_TYPE_CUDA);
+
+    if (UCC_OK != (status = ucc_mc_cuda_memcpy_kind_map(dst_mem, src_mem, &kind))) {
```

Comment:
I think we can skip finding the kind here. we can use cudaMemCpyDefault instead. which is recommended (https://docs.nvidia.com/cuda/cuda-runtime-api/group__CUDART__MEMORY.html#group__CUDART__MEMORY_1g85073372f776b4c4d5f89f7124b7bf79)

Reply:
@bureddy does that imply additional ptr kind detection inside libcuda? If we already know the ptr types do we want to do the checks again? Maybe it is cheap though, i'm not sure - what is your thought? @Sergei-Lebedev ? Also :  "However, cudaMemcpyDefault is only allowed on systems that support unified virtual addressing" - is this a limitation or we can ignore that ?

Reply:
I think driver knows the address ranges for memory its allocated. cuda unified virtual addressing is the default on all platforms for a long time now. I think we can ignore that.

Reply:
ok, changed to cudaMemcpyDefault